src/libtls/tls_hkdf.h

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (C) 2020 Pascal Knecht
+ * Copyright (C) 2020 Méline Sieber
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.  See <http://www.fsf.org/copyleft/gpl.txt>.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ */
+
+/**
+ * @defgroup tls_hkdf tls_hkdf
+ * @{ @ingroup libtls
+ */
+
+#ifndef TLS_HKDF_H_
+#define TLS_HKDF_H_
+
+#include <crypto/hashers/hasher.h>
+#include <utils/chunk.h>
+
+typedef struct tls_hkdf_t tls_hkdf_t;
+
+/**
+ * TLS HKDF labels
+ */
+enum tls_hkdf_labels_t {
+	TLS_HKDF_EXT_BINDER,
+	TLS_HKDF_RES_BINDER,
+	TLS_HKDF_C_E_TRAFFIC,
+	TLS_HKDF_E_EXP_MASTER,
+	TLS_HKDF_C_HS_TRAFFIC,
+	TLS_HKDF_S_HS_TRAFFIC,
+	TLS_HKDF_C_AP_TRAFFIC,
+	TLS_HKDF_S_AP_TRAFFIC,
+	TLS_HKDF_EXP_MASTER,
+	TLS_HKDF_RES_MASTER,
+};
+
+/**
+ * TLS HKDF helper functions.
+ */
+struct tls_hkdf_t {
+
+	/**
+	 * Set the (EC)DHE shared secret of this connection.
+	 *
+	 * @param shared_secret		input key material to use
+	 * @return 					TRUE if the ikm set successfully
+	 */
+	bool (*set_shared_secret)(tls_hkdf_t *this, chunk_t *shared_secret);
```

Comment:
The secret should not be passed as pointer. And as mentioned in the implementation, it should be clearly stated who is responsible for the passed secret.